lib/ui/painting.dart

```diff
@@ -2869,44 +2906,79 @@ class ImageFilter {
 
   final Float64List _data;
   final FilterQuality? _filterQuality;
+  final ColorFilter? _colorFilter;
+  final ImageFilter? _innerFilter;
+  final ImageFilter? _outerFilter;
   final int _type;
-  _ImageFilter? _nativeFilter;
+  late final _ImageFilter _nativeFilter = _makeNativeImageFilter();
 
   // The type of SkImageFilter class to create for Skia.
-  static const int _kTypeBlur = 0;   // MakeBlurFilter
-  static const int _kTypeMatrix = 1; // MakeMatrixFilterRowMajor255
+  static const int _kTypeComposed = 0;    // SkImageFilters::Compose
+  static const int _kTypeBlur = 1;        // MakeBlurFilter
+  static const int _kTypeMatrix = 2;      // MakeMatrixFilterRowMajor255
+  static const int _kTypeColorFilter = 3; // SkImageFilters::ColorFilter
 
   @override
   bool operator ==(Object other) {
     return other is ImageFilter
         && other._type == _type
         && _listEquals<double>(other._data, _data)
-        && other._filterQuality == _filterQuality;
+        && other._filterQuality == _filterQuality
+        && other._outerFilter == _outerFilter
+        && other._innerFilter == _innerFilter
+        && other._colorFilter == _colorFilter;
   }
 
-  _ImageFilter _toNativeImageFilter() => _nativeFilter ??= _makeNativeImageFilter();
+  _ImageFilter _toNativeImageFilter() => _nativeFilter;
 
   _ImageFilter _makeNativeImageFilter() {
     switch (_type) {
+      case _kTypeComposed:
+        return _ImageFilter.composed(this);
       case _kTypeBlur:
         return _ImageFilter.blur(this);
       case _kTypeMatrix:
         return _ImageFilter.matrix(this);
+      case _kTypeColorFilter:
+        return _ImageFilter.fromColorFilter(this);
       default:
         throw StateError('Unknown mode $_type for ImageFilter.');
     }
   }
 
   @override
-  int get hashCode => hashValues(_filterQuality, hashList(_data), _type);
+  int get hashCode => hashValues(_innerFilter, _outerFilter, _filterQuality, hashList(_data), _type, _colorFilter);
+
+  String? get _shortDescription {
+    switch (_type) {
+      case _kTypeComposed:
+        final String? outer = _outerFilter?._shortDescription;
+        final String? inner = _innerFilter?._shortDescription;
+        return (inner != null && outer != null) ? '$inner -> $outer' : null;
+      case _kTypeBlur:
+        return 'blur(${_data[0]}, ${_data[1]})';
+      case _kTypeMatrix:
+        return 'matrix($_data, $_filterQuality)';
+      case _kTypeColorFilter:
+        return '$_colorFilter';
+      default:
+        return null;
+    }
+  }
 
   @override
   String toString() {
     switch (_type) {
+      case _kTypeComposed:
+        return _shortDescription != null
+          ? 'ImageFilter.compose(from innermost to outermost: $_shortDescription)'
+          : 'Unknown ImageFilter type. This is an error. If you\'re seeing this, please file an issue at https://github.com/flutter/flutter/issues/new.';
```

Comment:
Why not just let it be `ImageFilter.compose(...: null)` as its own mystery to the debugging developer?

Reply:
(Still not sure why you can compose null filters...?)